### PR TITLE
#242 PersistentBuffer & BroadcastBuffer commit safety

### PR DESCRIPTION
**Pull request by anilgursel**
Thanks for your pull request.  Please review the following guidelines.

- [x] Title includes issue id.
- [x] Description of the change added.
- [x] Commits are squashed.
- [x] Tests added.
- [x] Documentation added/updated.
- [x] Also please review [CONTRIBUTING.md](https://github.com/paypal/squbs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: squbs-pattern/src/main/scala/org/squbs/pattern/stream/PersistentQueue.scala [hunk 1]
```diff
@@ -126,11 +130,34 @@ class PersistentQueue[T](config: QueueConfig, onCommitCallback: Int => Unit = (x
...
   def commit(outputPortId: Int, index: Long): Unit = {
+    verifyCommitOrder(outputPortId, index)
     if (!indexMounted) mountIndexFile()
     indexStore.writeLong(outputPortId << 3, index)
     onCommitCallback(outputPortId)
   }
 
+  private def verifyCommitOrder(outputPortId: Int, index: Long): Unit = {
+    val lastCommit = lastCommitIndex(outputPortId)
+    if(index == lastCommit + 1) lastCommitIndex(outputPortId) = index
```
Comment (anilgursel) [inline squbs-pattern/src/main/scala/org/squbs/pattern/stream/PersistentQueue.scala]: Do we actually want this if check even for `Lenient` use case?  If we would like to add a log, we should..  However, if log is not needed, then I can control this at the buffer level and use a different function..

File: squbs-pattern/src/main/scala/org/squbs/pattern/stream/PersistentQueue.scala [hunk 2]
```diff
@@ -126,11 +130,34 @@ class PersistentQueue[T](config: QueueConfig, onCommitCallback: Int => Unit = (x
...
+        lastCommitIndex(outputPortId) = index
+      } else {
+        config.commitOrderPolicy match {
+          case Lenient =>
+            logger.warn(s"Missing or out of order commits.  previous: ${lastCommitIndex(outputPortId)} latest: $index cycle: ${cycle(outputPortId)}")
+            lastCommitIndex(outputPortId) = index
+          case Strict =>
+            val msg = s"Missing or out of order commits.  previous: ${lastCommitIndex(outputPortId)} latest: $index cycle: ${cycle(outputPortId)}"
+            logger.error(msg)
+            throw new CommitOrderException(msg, lastCommitIndex(outputPortId), index, cycle(outputPortId))
```
Comment (anilgursel) [inline squbs-pattern/src/main/scala/org/squbs/pattern/stream/PersistentQueue.scala]: Not closing the queue here as the `SupervisionDecider` might resume the stream.

File: squbs-pattern/src/main/scala/org/squbs/pattern/stream/PersistentQueue.scala (hunk 1 above)
Comment (akara) [inline squbs-pattern/src/main/scala/org/squbs/pattern/stream/PersistentQueue.scala]: Looking at the code, it costs pretty much one single branch and one non-concurrent assignment. I think that should not be an issue on modern processors.

File: squbs-pattern/src/main/scala/org/squbs/pattern/stream/PersistentQueue.scala (hunk 2 above)
Comment (akara) [inline squbs-pattern/src/main/scala/org/squbs/pattern/stream/PersistentQueue.scala]: Thanks for the clarification. Perhaps you want to stick this into a comment on the code, too. Future maintainers won't make the mistake of closing the queue. They usually don't look at your github comments on the original commits before they make the mistake 😄

File: squbs-pattern/src/test/scala/org/squbs/pattern/stream/BroadcastBufferCommitOrderSpec.scala [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    import util._
+    val buffer = new BroadcastBuffer[Int](ConfigFactory.parseString("commit-order-policy = strict").withFallback(config))
+    val streamGraph = RunnableGraph.fromGraph(GraphDSL.create(flowCounter) { implicit builder =>
+      sink =>
+        import GraphDSL.Implicits._
+        val commit = buffer.commit // makes a dummy flow if autocommit is set to false
+      val bcBuffer = builder.add(buffer.async)
+        val mr = builder.add(merge)
+        in ~> bcBuffer ~> filterARandomElement ~> commit ~> mr ~> sink
+        bcBuffer ~> commit ~> mr
```
Comment (akara) [inline squbs-pattern/src/test/scala/org/squbs/pattern/stream/BroadcastBufferCommitOrderSpec.scala]: Nice simulation of the failure!

**Pull request merged**